### PR TITLE
Add documentation and e2e for BackupPolicy API

### PR DESCRIPTION
**Pull request by rfashwall-anynines**
# Short Description

- Update Readme for `a8s-backup-manager` to include `BackupPolicy` API
- Add Sample manifest of BackupPolicy
# Details

# Note

WARNING: Only users listed in the CODEOWNERS file can approve PRs!

# Checks

- [ ] Documentation has been adjusted
- [ ] Architectural decisions have been documented
- [ ] Changelog has been updated
- [ ] PR is approved by a code owner
- [ ] Manifests are updated
- [ ] Commit message adheres to our [guideline](https://anynines.atlassian.net/wiki/spaces/DS/pages/2423193626/Version+Control+Workflow)

### CODE REVIEW

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]:
```suggestion
The `BackupPolicy` controller manages the scheduling and lifecycle of `Backup` resources in a Kubernetes cluster. It ensures backups are created according to a defined schedule and retains or deletes old backups based on a specified retention policy.
```

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md:1]: I would suggest installing https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint to provide markdown linting to help suggest small improvements to formatting here.

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]: What do you mean `varias schedule types`. Could you clarify this sentence.

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]:
```suggestion
Allows defining retention policies based on either the number of backups to retain or the maximum age of backups. You can retain backups for a specified number or define a maximum age for backups (e.g., 7 days). Backups older than this period will be deleted. This age is calculated based on the creation timestamp of the backup. It supports multiple duration types: `hours`, `days`, `months`, and `years`.
```

Comment (mkfdoherty) [reply docs/application-developers/backuppolicy.md]: We might want to use period rather than age throughout. I will leave this to your discretion.

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]:
```suggestion
The ScheduleConfiguration defines the schedule type to be used. We currently support Cron configuration, which is the most common method for creating schedules. More types may be supported in the future based on user needs.

The configuration allows you to specify a timezone for calculating the schedule. The default is `UTC`.
```

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]:
```suggestion
The RetentionConfiguration defines the policy for retaining and deleting old backups. It ensures that backups are kept for a specified duration or up to a specified number of backups.
```

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]:
```suggestion
With retention by count, you define the maximum number of backups to retain. Backups exceeding this limit are deleted, starting with the oldest ones (based on creation timestamp).
```

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]: https://github.com/anynines/a8s-backup-manager/pull/83/files#r1687423156

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]:
```suggestion
In the above example, the controller fetches the backups created by the policy and checks their creationTimestamp against a 15-day window. If the timestamp falls outside this window (i.e., the backup is older than 15 days), it marks the backup for deletion.
```

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]: I don't think it makes sense to include this ticket in the user facing documentation. This could be included in the component repository for developers.

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]:
```suggestion
The newly created policy will create a backup immediately. A backup is created if there are no backups or if the scheduled time has elapsed.
```

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]:
```suggestion
After creating a policy, you can disable it, which allows you to keep it instead of just deleting the resource. However, updating the status and reconciliation will not be possible until it is enabled again. All policies created are enabled by default.
```

File: docs/application-developers/backuppolicy.md
Comment (mkfdoherty) [inline docs/application-developers/backuppolicy.md]:
```suggestion
The backup created will have a generated name with the convention backup-, where each backup created will be labeled with the key a8s.a9s/policy and the value as the policy name.
```

**Review by mkfdoherty — CHANGES_REQUESTED** [review]

File: test/go.mod
Comment (lhaendler) [inline test/go.mod]: Issue: This will need to be removed before merging

File: CHANGELOG.md
Comment (lhaendler) [inline CHANGELOG.md]:
```suggestion
* BackupPolicy to manage backups supporting scheduling and retention configuration. 
```

File: CHANGELOG.md
Comment (lhaendler) [inline CHANGELOG.md]: I think this section would be unclear to a user.

I would expect an explanation of what the new feature does. For example: 
> BackupPolicy feature: Create new backups based on a schedule. You can now configure a schedule for each Postgresql instance's backups and when the backups will be cleaned up.

Please remove the mentions of end-to-end tests. Users are not concerned about that.

File: docs/application-developers/backuppolicy.md
Comment (lhaendler) [inline docs/application-developers/backuppolicy.md]: This line is too long. Please split up too long lines (above 100 characters)

File: docs/application-developers/backuppolicy.md
Comment (lhaendler) [inline docs/application-developers/backuppolicy.md]: Suggestion: Try formulating documentation in an active mood. I'm using this paragraph as an example, but my suggestion applies throughout.

> The controller(/a8s framework) will delete Backups above this limit, starting with the oldest ones.

**Review by lhaendler — CHANGES_REQUESTED** [review]

File: docs/application-developers/backuppolicy.md
Comment (lhaendler) [inline docs/application-developers/backuppolicy.md]: Suggestion: 
```suggestion
Once you create the policy, the controller will create a backup immediately. The controller creates a backup if there are no backups or if the scheduled time has elapsed.
```

File: test/e2e/backuppolicy/backuppolicy_test.go
Comment (lhaendler) [inline test/e2e/backuppolicy/backuppolicy_test.go]: Did you mean
```suggestion
			By("Create existing backups", func() {
```
?

File: test/e2e/backuppolicy/backuppolicy_test.go
Comment (lhaendler) [inline test/e2e/backuppolicy/backuppolicy_test.go]: **nit**: this does not read very well. GInkgo uses these blocks of natural language so that sentences can be formed. Try to do that where possible.
```suggestion
			By("deleting backups over the requested count", func() {
```

File: test/e2e/backuppolicy/backuppolicy_test.go
Comment (lhaendler) [inline test/e2e/backuppolicy/backuppolicy_test.go:199]: **Question**: Shouldn't this be wrapped in an eventually? The backups are deleted by the controller asynchronously, right?

File: test/framework/backuppolicy/backuppolicy.go
Comment (lhaendler) [inline test/framework/backuppolicy/backuppolicy.go]:
```suggestion
func MockExistingBackups(p *v1beta3.BackupPolicy, dsi runtimeClient.Object) []v1beta3.Backup {
```
Are `existingBackups` test data?

File: test/framework/backuppolicy/backuppolicy.go
Comment (lhaendler) [inline test/framework/backuppolicy/backuppolicy.go]:
```suggestion
func CheckBackupsCount(ctx context.Context, p *v1beta3.BackupPolicy, c runtimeClient.Client, expectedLength int) *v1beta3.BackupList {
```

File: test/framework/backuppolicy/backuppolicy.go
Comment (lhaendler) [inline test/framework/backuppolicy/backuppolicy.go]: I would prefer having the eventually in the test code, instead of in the helper. This makes the test behavior more apparent at first glance.

File: test/framework/backuppolicy/backuppolicy.go
Comment (lhaendler) [inline test/framework/backuppolicy/backuppolicy.go:134]: Question: shouldn't the backup policy clean up the backups? Why do we need a utility function in the test library? Do we test that all backups are cleaned up when the policy is deleted?

**Review by lhaendler — COMMENTED** [review]
Many lines in the markdown files still exceed the team's limit of 100 characters. Please fix this by breaking them up.

File: test/framework/backuppolicy/backuppolicy.go
Comment (rfashwall-anynines) [reply test/framework/backuppolicy/backuppolicy.go:134]: @lhaendler backups will be deleted if retentionConfig is defined, we need it to clean the remaining backups also.
Now if the policy is deleted, no cascading deletion to the backups

File: test/e2e/backuppolicy/backuppolicy_test.go
Comment (mkfdoherty) [inline test/e2e/backuppolicy/backuppolicy_test.go:26]: The API is undefined since it does not exist in the version of `github.com/anynines/a8s-backup-manager` referenced in `go.mod`. We need to reference a release, commit or branch in `github.com/anynines/a8s-backup-manager` that includes the API for the type to be defined.

File: test/e2e/backuppolicy/backuppolicy_test.go
Comment (mkfdoherty) [inline test/e2e/backuppolicy/backuppolicy_test.go]: This gives a compiler error as `&backupPolicy` is an UnaddressableOperand.

File: test/e2e/backuppolicy/backuppolicy_test.go
Comment (mkfdoherty) [inline test/e2e/backuppolicy/backuppolicy_test.go]: `p` is not defined here.

File: test/framework/backuppolicy/backuppolicy.go
Comment (mkfdoherty) [reply test/framework/backuppolicy/backuppolicy.go:134]: @lhaendler Would you expect the deletion of the BackupPolicy to result in the deletion of the backups?

File: test/e2e/backuppolicy/backuppolicy_test.go
Comment (mkfdoherty) [reply test/e2e/backuppolicy/backuppolicy_test.go:199]: This still is not asynchronous and will likely never pass as a result. Could you test with and without the eventually?

File: test/e2e/backuppolicy/backuppolicy_test.go
Comment (mkfdoherty) [inline test/e2e/backuppolicy/backuppolicy_test.go:169]: The idea is to not mock with the e2e tests. Why don't we just create existing backups?

File: test/e2e/backuppolicy/backuppolicy_test.go
Comment (mkfdoherty) [inline test/e2e/backuppolicy/backuppolicy_test.go:73]: We don't just want to check that we pass API validation and can create the resource but also that it creates successfully via an expected status. Could we do something like that here?

File: examples/backuppolicy.yaml
Comment (mkfdoherty) [inline examples/backuppolicy.yaml]: `Postgresql` will fail as the a8s-backup-controller included in a8s-deployment expects `PostgreSQL`.

**Review by mkfdoherty — CHANGES_REQUESTED** [review]

File: test/framework/backuppolicy/backuppolicy.go
Comment (lhaendler) [reply test/framework/backuppolicy/backuppolicy.go:134]: @rfashwall-anynines and I have discussed this and we have come to the conclusion that this could be configurable, a ticket has been created to handle it in the future

**Review by mkfdoherty — APPROVED** [review]

**Review by lhaendler — APPROVED** [review]

**Pull request merged**
